### DISCUSSION TITLE
It might be useful for people to be able to inject their own security middleware into the api routes

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1612] [Show and tell]
@ganigeorgiev 

I created a casbin authentication middleware for my custom routes.  I'm thinking that this could be useful to someone for the built-in Pocketbase API as well.  For example, intensive operations could be blocked by simply not permitting it via a line in the casbin `policy.csv` file.  For example, maybe not allowed an expand for all users, or a certain group of users.  This would allow people to keep the collection API Rules fairly simple for referential integrity concerns only, and not cluttering it up with group permission checks.

By using a casbin middleware they could either leave out the annoying `request.authId != ''` stuff because this would fall under their `anon` user the the middleware.  It could then be allowed, or not.  Additionally, if something was very important then they could have two layers of security, one in the casbin rule prohibiting it, and also in the API rule.

Also, if casbin were ever integrated directly into Pocketbase, then each collection could have a toggle to use it or not, and then the rules could be added directly into a textarea.  Also by integrating it into Pocketbase then you would be able to offer the release binary using people the option of having an extremely robust and lightweight way to have a deny first, or allow first security system.  Due to the policy and model definitions just being loadable text, the only issue is what it means to belong to a group.  I think that you could provide it to either be a direct field on the auth record or via a direct field on a related collection.

Even if this is too much to think about all at once, it would be nice if you could allow developers to add their own middleware to the built-in collections api.

# General library code

```golang
// pkg/auth/auth.go

type Settings struct {
	SigninPath string
	GroupField string
	RootGroup  string
	AnonGroup  string
}

type Auth struct {
	enforcer *casbin.Enforcer
	settings Settings
}

func New(enforcer *casbin.Enforcer, settings Settings) (*Auth, error) {
	return &Auth{
			enforcer: enforcer,
			settings: settings,
		},
		nil
}

func (a *Auth) Allow(c echo.Context) bool {
	sub := "@" + a.group(c)
	obj := c.Path()
	act := c.Request().Method

	alw, err := a.enforcer.Enforce(sub, obj, act)
	return alw && err == nil
}

func (a *Auth) AllowForceSignin(c echo.Context) (alw bool, err error) {
	alw = a.Allow(c)
	if !alw && User(c) == nil {
		err = c.Redirect(http.StatusSeeOther, a.settings.SigninPath)
	}
	return
}

func (a *Auth) group(c echo.Context) string {
	if adm := Admin(c); adm != nil {
		return a.settings.RootGroup
	}

	if u := User(c); u != nil {
		return u.GetString(a.settings.GroupField)
	}

	return a.settings.AnonGroup
}
```

```golang
// pkg/auth/context.go

func User(c echo.Context) *models.Record {
	record, _ := c.Get(apis.ContextAuthRecordKey).(*models.Record)
	return record
}

func Admin(c echo.Context) *models.Admin {
	admin, _ := c.Get(apis.ContextAdminKey).(*models.Admin)
	return admin
}
```

The logic of the RestrictRoute middeware is that if checks the echo route path string (with variable placeholder syntax) against what you set in `policy.csv`.  I made it so that if it fails and you're logged in then it hard rejects the request, otherwise it redirects to whatever you set for the signin path `/signin`.

```golang
// pkg/auth/middleware.go

func RestrictRoute(auth *Auth) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			if alw, err := auth.AllowForceSignin(c); !alw || err != nil {
				return err
			}

			return next(c)
		}
	}
}
```

# Code  that utilizes the internal "pkg" code

```golang
// cmd/web/auth/auth.go

func New(cfg *config.Config) (*auth.Auth, error) {
	e, err := newEnforcer()
	if err != nil {
		return nil, err
	}

	return auth.New(e, auth.Settings{
		SigninPath: "/signin",
		GroupField: "group",
		RootGroup:  "root",
		AnonGroup:  "anon",
	})
}
```

```golang
// cmd/web/auth/enforcer.go

//go:embed model.conf policy.csv
var authFS embed.FS

func newEnforcer() (*casbin.Enforcer, error) {
	model, _ := casbin_helper.NewModelFromFilesystem(authFS, "model.conf")
	policies := casbin_helper.NewFilesystemAdapter(authFS, "policy.csv")
	return casbin.NewEnforcer(model, policies)
}
```

```ini
# cmd/web/auth/model.conf

[request_definition]
r = sub, obj, act

[policy_definition]
p = obj, sub, act

[role_definition]
g = _, _

[policy_effect]
e = some(where (p.eft == allow))

[matchers]
m = g(r.sub, p.sub) && keyMatch2(r.obj, p.obj) && regexMatch(r.act, p.act) || r.sub == "@root"
```

```csv
# cmd/web/auth/policy.csv

# site controller
p, /site/create, @normal, GET

# home controller
p, /home, @normal, GET

# appi controller
p, /appi/site/is-name-available, @normal, POST
p, /appi/site/create, @normal, POST


######## GROUPS ########
g, @normal, @anon
g, @admin, @normal
```

# Route setup

The idea here is that anything that is attached to the `rgRestrict` router group is going to have it's echo route string validated with the casbin enforcer.  Casbin is ideal here because you can change the rules in `model.conf` (embedded into the binary) in wild creative ways.

```golang
func Init(env *web.Env, r *echo.Echo) {
	rgPublic, rgAssets, rgRestrict := routeGroups(env, r)

	bindPublic(env, rgPublic)
	bindSession(env, rgPublic)
	bindAccount(env, rgPublic)

	appi.Bind(env, rgRestrict)
	bindSite(env, rgRestrict)
	bindHome(env, rgRestrict)

	bindAssets(env, rgAssets)
}

func routeGroups(env *web.Env, r *echo.Echo) (*echo.Group, *echo.Group, *echo.Group) {
	rgPublic := r.Group("")
	rgPublic.Use(
		echo_middleware.Recover(),
		echo_middleware.Secure(),
		auth.FromUserCookie(env.Dao, env.Settings),
	)

	rgRestrict := rgPublic.Group("")
	rgRestrict.Use(
		auth.RestrictRoute(env.Auth),
	)

	rgAssets := r.Group("")
	rgAssets.Use(
		echo_middleware.Secure(),
	)

	return rgPublic, rgAssets, rgRestrict
}
```

**Comment by hsluoyz** [comment]
@dschissler it's amazing! Can you put all the code into a standalone repo?

**Reply by dschissler** [reply]
@hsluoyz I think that I could.  I think that I refactored my code a few times after I posted this.  I'm considering to make a repo like you requested.

**Reply by epuerta9** [reply]
please tell me you got around to making a repo for this? I'm just visiting pocketbase and I previously had managed Authz with Casbin so this is still incredibly relevant
